workspaces/pingidentity/plugins/catalog-backend-module-pingidentity/src/lib/read.ts

```diff
@@ -0,0 +1,174 @@
+  client: PingIdentityClient,
+  groups: GroupEntity[],
+  groupMembersMap: Map<string, Set<string>>,
+  userTransformer?: UserTransformer
+): Promise<UserEntity[]> => {
+  const transformer = userTransformer ?? defaultUserTransformer;
+  const pingIdentityUsers: PingIdentityUser[] = await client.getUsers();
+  const transformedUsers: (UserEntity | undefined)[] = await Promise.all(pingIdentityUsers.map(async (user: any) => {
+    const userLocation = getEntityLocation(client.getConfig(), 'users', user.id);
+    return await transformer({
+      apiVersion: 'backstage.io/v1beta1',
+      kind: 'User',
+      metadata: {
+        name: user.username!,
+        annotations: {
+          [PING_IDENTITY_ID_ANNOTATION]: user.id!,
+          [ANNOTATION_LOCATION]: userLocation,
+          [ANNOTATION_ORIGIN_LOCATION]: userLocation,
+        },
+      },
+      spec: {
+        profile: {
+          email: user.email,
+          ...(user.name.given || user.name.family
+            ? {
+              displayName: [user.name.given, user.name.family]
+                .filter(Boolean)
+                .join(' '),
+            }
+            : {}),
+        },
+        memberOf: findGroupMemberships(user.id, groups, groupMembersMap)
+      }
+    }, user, client.getConfig().envId, groups);
+  }));
+  return transformedUsers.filter(user => user !== undefined) as UserEntity[];
+}
+
+/**
+ * Returns a parsed list of all groups in the Ping Identity environment
+ *
+ * @param client - The `PingIdentityClient`
+ * @param groupMembersMap - Maps group ID to all user IDs that belong in that group
+ * @param groupTransformer - Optional group transformer method
+ * 
+ * @returns a parsed list of all groups fetched from Ping Identity of type `GroupEntity`
+ */
+const parsePingIdentityGroups = async (
+  client: PingIdentityClient,
+  groupMembersMap: Map<string, Set<string>>,
+  groupTransformer?: GroupTransformer
+): Promise<GroupEntity[]> => {
+  const transformer = groupTransformer ?? defaultGroupTransformer;
+  const pingIdentityGroups: PingIdentityGroup[] = await client.getGroups();
+  const transformedGroups: (GroupEntity | undefined)[] = await Promise.all(pingIdentityGroups.map(async (group: any) => {
+    const groupLocation = getEntityLocation(client.getConfig(), 'groups', group.id);
+    // add users in group to group membership map
+    groupMembersMap.set(group.id, new Set(await client.getUsersInGroup(group.id)));
+    return await transformer({
+      apiVersion: 'backstage.io/v1beta1',
+      kind: 'Group',
+      metadata: {
+        name: group.name!,
+        annotations: {
+          [PING_IDENTITY_ID_ANNOTATION]: group.id!,
+          [ANNOTATION_LOCATION]: groupLocation,
+          [ANNOTATION_ORIGIN_LOCATION]: groupLocation,
+        },
+        description: group.description
+      },
+      spec: {
+        type: 'group',
+        profile: {
+          displayName: group.name!,
+        },
+        children: [],
```

Comment:
yes we do, there is a test case for this [here](https://github.com/backstage/community-plugins/blob/4da466112dce407be2c594275c30eab89bc7b798/workspaces/pingidentity/plugins/catalog-backend-module-pingidentity/src/lib/read.test.ts#L222-L278).

Reply:
Are we not populating `children` to contain subgroups?

Reply:
I skip this because Backstage will establish the relation as long as the child group has the parent group set and the UI reflects this properly.

In the entity yaml of the parent it looks like:
```
relations:
  - type: parentOf
    targetRef: group:default/group_three
    target:
      kind: group
      namespace: default
      name: group_three
spec:
  type: group
  profile:
    displayName: group one
  children: [] //empty
  members: []
```

In the child group yaml:
```
relations:
  - type: childOf
    targetRef: group:default/group_one
    target:
      kind: group
      namespace: default
      name: group_one
spec:
  type: group
  profile:
    displayName: group three
  children: []
  parent: group_one //defined
  members: []
```
Imo this should suffice, what do you think?

Reply:
> as long as the child group has the parent group set 
Is this set automatically or does that relationship need to be explicitly set?

Also, under what circumstances would the `children` field be used?  If it's redundant then can it be removed?

Reply:
> Is this set automatically or does that relationship need to be explicitly set?

I do have to set this explicitly [here](https://github.com/backstage/community-plugins/blob/f48dbf128ea50dcec02f2776637c1e8b3033d036/workspaces/pingidentity/plugins/catalog-backend-module-pingidentity/src/lib/read.ts#L185-L198)

> Also, under what circumstances would the children field be used? If it's redundant then can it be removed?
Good point! It's never used, same with `group.spec.member` since I set `user.spec.memberOf`. I'll go ahead and remove these fields.

Reply:
If `group.spec.children` is mandatory then is this approach anti-pattern?  Just wondering if `children` might be used somewhere else in backstage? I can see how it can save processing time if children refs are populated in the spec.

Reply:
No it's OK. People are free to choose whether to set parent or children (or both). The symmetric pair of relations generated between them will be exactly the same in either case - and consumers of the data are meant to look at relations as the authority, not the spec. To the point that we've considered even deleting the spec fields after processing so they aren't visible in the final entity, to avoid mistakes :)

Reply:
The reason that both options exist in the first place is just convenience for data producers. Some data sources already have a parent pointer redily at hand but would have to take a performance hit explicitly forming a children array query, or vice versa. So it's just a nice way for an entity provider to pöonk down what it has at hand, and things just working out.

Reply:
@freben Got it, thanks so much for the info! I'll go ahead and resolve this then.